### PR TITLE
Implement snapshotting

### PR DESCRIPTION
**Pull request by malfarhan7**
Implement snapshotting to save snapshots of the forward wavefield used to compute the gradient to reduce memory usage.

### CODE REVIEW

File: devitofwi/waveengine/acoustic.py [hunk 1]
```diff
@@ -417,7 +424,8 @@ def _loss_grad_oneshot(self, vp, src, solver, isrc,
 
         """
         # Compute synthetic data and full forward wavefield u0
-        adjsrc, u0, _ = solver.forward(vp=vp, save=True, src=src, autotune=True)
+        adjsrc, u0, usnaps, _ = solver.forward(vp=vp, save=True, src=src, autotune=True,
```
Comment (mrava87) [inline devitofwi/waveengine/acoustic.py]: I think here we are still saving `u0`.

I did a test adding two prints:

```
print('Save', True if self.factor is None else False)
adjsrc, u0, usnaps, _ = solver.forward(vp=vp, save=True, src=src, autotune=True, factor=self.factor)
print('u0', u0, u0.shape, 'usnaps', usnaps)        
```
and I get:
```
Save False
u0 u(time, x, y) (4107, 641, 261) usnaps usnaps(t_sub, x, y)
```

Now if i change the code to:
```
print('Save', True if self.factor is None else False)
adjsrc, u0, usnaps, _ = solver.forward(vp=vp, save=True if self.factor is None else False, 
                                                                 src=src, autotune=True, factor=self.factor)
print('u0', u0, u0.shape, 'usnaps', usnaps)        
```

I get:
```
Save False
u0 u(t, x, y) (3, 641, 261) usnaps usnaps(t_sub, x, y)
```

**Review by mrava87 — COMMENTED** [review]
@malfarhan7 this looks great, well done! 

I have only found a place in the code where I was wondering if we are still saving `u` despite not needing it... see my detailed comment.

Other than that, I made a few small changes to the notebook (including a plot of the wavelet and its spectrum as a way to help choosing an appropriate `factor`), and added a few scripts with just the gradient computations to memory profile. 

I will push the code if you agree with the major change I suggest :)

**Comment by mrava87** [comment]
@malfarhan7 ready to be merged for me. Agree?

**Review by malfarhan7 — COMMENTED** [review]
Look good to me!

**Pull request merged**